File: docs/examples/python/dae_multiple_shooting.py
#
#     MIT No Attribution
#
#     Copyright 2023 Joel Andersson, Joris Gillis, Moritz Diehl, KU Leuven.
#
#     Permission is hereby granted, free of charge, to any person obtaining a copy of this
#     software and associated documentation files (the "Software"), to deal in the Software
#     without restriction, including without limitation the rights to use, copy, modify,
#     merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
#     permit persons to whom the Software is furnished to do so.
#
#     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
#     INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
#     PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
#     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
#     OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
#     SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#
#
from casadi import *

"""
Solves the following optimal control problem (OCP) in differential-algebraic
equations (DAE)

minimize     integral_{t=0}^{10} x0^2 + x1^2 + u^2  dt
x0,x1,z,u

subject to   dot(x0) == z*x0-x1+u     \
             dot(x1) == x0             }  for 0 <= t <= 10
                   0 == x1^2 + z - 1  /
             x0(t=0) == 0
             x1(t=0) == 1
             x0(t=10) == 0
             x1(t=10) == 0
             -0.75 <= u <= 1  for 0 <= t <= 10

The method used is direct multiple shooting.

Joel Andersson, 2015
"""

# Declare variables
x0 = SX.sym('x0')
x1 = SX.sym('x1')
x = vertcat(x0, x1) # Differential states
z = SX.sym('z')       # Algebraic variable
u = SX.sym('u')       # Control

# Differential equation
f_x = vertcat(z*x0-x1+u, x0)

# Algebraic equation
f_z = x1**2 + z - 1

# Lagrange cost term (quadrature)
f_q = x0**2 + x1**2 + u**2

# Create an integrator
dae = {'x':x, 'z':z, 'p':u, 'ode':f_x, 'alg':f_z, 'quad':f_q}
# interval length 0.5s
I = integrator('I', 'idas', dae, 0, 0.5)

# Number of intervals
nk = 20

# Start with an empty NLP
w = []   # List of variables
lbw = [] # Lower bounds on w
ubw = [] # Upper bounds on w
G = []   # Constraints
J = 0    # Cost function

# Initial conditions
Xk = MX.sym('X0', 2)
w.append(Xk)
lbw += [ 0, 1 ]
ubw += [ 0, 1 ]

# Loop over all intervals
for k in range(nk):
  # Local control
  Uk = MX.sym('U'+str(k))
  w.append(Uk)
  lbw += [-0.75]
  ubw += [ 1.00]

  # Call integrator function
  Ik = I(x0=Xk, p=Uk)
  Xk = Ik['xf']
  J = J + Ik['qf'] # Sum quadratures

  # "Lift" the variable
  X_prev = Xk
  Xk = MX.sym('X'+str(k+1), 2)
  w.append(Xk)
  lbw += [-inf, -inf]
  ubw += [ inf,  inf]
  G.append(X_prev - Xk)

# Allocate an NLP solver
nlp = {'x':vertcat(*w), 'f':J, 'g':vertcat(*G)}
opts = {'ipopt.linear_solver':'ma27'}
solver = nlpsol('solver', 'ipopt', nlp, opts)

# Pass bounds, initial guess and solve NLP
sol = solver(lbx = lbw, # Lower variable bound
             ubx = ubw,  # Upper variable bound
             lbg = 0.0,  # Lower constraint bound
             ubg = 0.0,  # Upper constraint bound
             x0  = 0.0) # Initial guess

# Plot the results
import matplotlib.pyplot as plt
plt.figure(1)
plt.clf()
plt.plot(linspace(0., 10., nk+1), sol['x'][0::3],'--')
plt.plot(linspace(0., 10., nk+1), sol['x'][1::3],'-')
plt.plot(linspace(0., 10., nk), sol['x'][2::3],'-.')
plt.title('Van der Pol optimization - multiple shooting')
plt.xlabel('time')
plt.legend(['x0 trajectory','x1 trajectory','u trajectory'])
plt.grid()
plt.show()

File: docs/examples/julia/dae_multiple_shooting.jl
#
#     MIT No Attribution
#
#     Copyright (C) 2010-2023 Joel Andersson, Joris Gillis, Moritz Diehl, KU Leuven.
#
#     Permission is hereby granted, free of charge, to any person obtaining a copy of this
#     software and associated documentation files (the "Software"), to deal in the Software
#     without restriction, including without limitation the rights to use, copy, modify,
#     merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
#     permit persons to whom the Software is furnished to do so.
#
#     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
#     INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
#     PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
#     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
#     OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
#     SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#
# Direct multiple shooting for the same DAE optimal control problem as
# dae_single_shooting. "collocation" stands in for the python "idas" integrator.
const _jl_dir = get(ENV, "CASADI_JL",
    joinpath(@__DIR__, "..", "..", "..", "julia-proto", "casadi-contact"))
include(joinpath(_jl_dir, "CasADiNative.jl"))
import .CasADiNative as ca

# Declare variables
x0 = ca.SX.sym("x0")
x1 = ca.SX.sym("x1")
x = ca.vertcat(x0, x1)   # Differential states
z = ca.SX.sym("z")      # Algebraic variable
u = ca.SX.sym("u")      # Control

# Differential equation
f_x = ca.vertcat(z * x0 - x1 + u, x0)

# Algebraic equation
f_z = x1^2 + z - 1

# Lagrange cost term (quadrature)
f_q = x0^2 + x1^2 + u^2

# Create an integrator (interval length 0.5s)
dae = Dict("x" => x, "z" => z, "p" => u, "ode" => f_x, "alg" => f_z, "quad" => f_q)
I = ca.integrator("I", "collocation", dae, 0.0, 0.5)

# Number of intervals
nk = 20

function main()
    # Start with an empty NLP
    w = ca.MX[]              # List of variables
    lbw = Float64[]       # Lower bounds on w
    ubw = Float64[]       # Upper bounds on w
    G = ca.MX[]              # Constraints
    J = ca.MX(ca.DM(0.0))   # Cost function

    # Initial conditions
    Xk = ca.MX.sym("X0", 2)
    push!(w, Xk)
    append!(lbw, [0, 1]); append!(ubw, [0, 1])

    # Loop over all intervals
    for k in 0:nk - 1
        # Local control
        Uk = ca.MX.sym("U$k")
        push!(w, Uk)
        push!(lbw, -0.75); push!(ubw, 1.00)

        # Call integrator function
        Ik = I(x0 = Xk, p = Uk)
        Xk = Ik["xf"]
        J = J + Ik["qf"]   # Sum quadratures

        # "Lift" the variable
        X_prev = Xk
        Xk = ca.MX.sym("X$(k + 1)", 2)
        push!(w, Xk)
        append!(lbw, [-Inf, -Inf]); append!(ubw, [Inf, Inf])
        push!(G, X_prev - Xk)
    end

    # Allocate an NLP solver
    nlp = Dict("x" => ca.vertcat(w), "f" => J, "g" => ca.vertcat(G))
    solver = ca.nlpsol("solver", "ipopt", nlp)

    # Pass bounds, initial guess and solve NLP
    sol = solver(lbx = lbw, ubx = ubw, lbg = 0.0, ubg = 0.0, x0 = 0.0)

    println("-----")
    println("objective at solution = ", sol["f"])
    println("x0 trajectory = ", sol["x"][1:3:end])
    println("x1 trajectory = ", sol["x"][2:3:end])
end

main()
